src/Framework/Constraint/Count.php

```diff
@@ -65,7 +66,7 @@ protected function getCountOf($other)
                 return $this->getCountOfGenerator($iterator);
             }
 
-            $key   = $iterator->key();
+            $key   = ($iterator instanceof Iterator) ? $iterator->key() : null;
```

Comment:
How do you propose testing this in userland if Traversable cannot be implemented directly?

Given the current implementation, I could exercise the `! $iterator instanceof Iterator` condition by having an IteratorAggregate instance return a second IteratorAggregate; however, that makes an assumption that the earlier logic that resolves an IteratorAggregate will always stop after one level of resolution.

I'd also like to discuss the two outstanding questions in the OP before making further changes. It would be premature to invest time in a test case until we establish what the correct fix should be.